### PR TITLE
Add MedCon command

### LINKED ISSUE #86: Add command to update patient's medical condition
So that doctor can keep track of patient's diagnosis

### PR DESCRIPTION
**Pull request by TechRaj**
Fixes #86

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+     * @return A {@link AddMedConCommand} object containing the parsed NRIC and medical conditon.
+     * @throws ParseException if the user input does not conform to the expected format or
+     *         if the NRIC or medical condition is not provided.
+     */
+    public AddMedConCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        String trimmedArgs = args.trim();
+        ArgumentMultimap argMultimap = ArgumentTokenizer.tokenize(args,
+                PREFIX_NRIC, PREFIX_MEDCON);
+        System.out.println(trimmedArgs);
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java:30]: Use a debugger over print statements as you might forget to remove these print statements. Please remove it in the next commit.

File: src/main/java/seedu/address/logic/commands/AddMedConCommand.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import seedu.address.logic.commands.exceptions.CommandException;
+import seedu.address.model.Model;
+import seedu.address.model.person.Nric;
+import seedu.address.model.person.Person;
+import seedu.address.model.person.MedCon;
+
+/**
+ * Command to assign a medical condition to a patient in the address book.
+ */
+public class AddMedConCommand extends Command {
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/commands/AddMedConCommand.java:18]: Just a nit: use full name over short form naming for consistency.

File: src/main/java/seedu/address/logic/commands/AddMedConCommand.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    public static final String MESSAGE_ARGUMENTS = "NRIC: %1$d, MedCon: %2$s";
+
+    private final Nric nric;
+    private final MedCon medCon;
+
+    /**
+     * @param nric of the patient to assign the medical condition to
+     * @param medCon of the patient
+     */
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/commands/AddMedConCommand.java:39]: Javadocs should have a short summary. Please use punctuation after every param.

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 4]
```diff
@@ -116,9 +117,12 @@ private static Person createEditedPerson(Person personToEdit, EditPersonDescript
         Set<Tag> updatedTags = editPersonDescriptor.getTags().orElse(personToEdit.getTags());
         Priority updatedPriority = personToEdit.getPriority();
         Set<Appointment> updatedAppointments = editPersonDescriptor.getAppointments()
-            .orElse(personToEdit.getAppointments());
+                                                                   .orElse(personToEdit.getAppointments());
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:120]: Should be 8 spacing from parent.

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 5]
```diff
@@ -246,4 +247,22 @@ public static Set<Appointment> parseAppointments(Collection<String> appointments
...
+     * @throws ParseException if the provided string does not conform to the expected
+     *         format or is invalid as per the priority constraints defined in the
+     *         {@link Priority} class.
+     */
+    public static MedCon parseMedCon(String medConStr) throws ParseException {
+        requireNonNull(medConStr);
+        String trimmedMedCon = medConStr.trim();
+//        if (!MedCon.isValidMedCon(trimmedPriority)) {
+//            throw new ParseException(Priority.MESSAGE_CONSTRAINTS);
+//        }
```
Comment (hweinian) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java:265]: Why is code commented out?

File: src/main/java/seedu/address/model/person/MedCon.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ * Represents a Person's remark in the address book.
+ * Guarantees: immutable; is always valid
+ */
+public class MedCon {
+
+    public final String value;
+
+    public MedCon(String medCon) {
+        requireNonNull(medCon);
+        value = medCon;
```
Comment (hweinian) [inline src/main/java/seedu/address/model/person/MedCon.java:14]: Could add javadocs

File: src/main/java/seedu/address/model/person/Person.java [hunk 7]
```diff
@@ -133,14 +139,16 @@ public boolean equals(Object other) {
                 && gender.equals(otherPerson.gender)
                 && address.equals(otherPerson.address)
                 && tags.equals(otherPerson.tags)
+                && appointments.equals(otherPerson.appointments)
+                && medCon.equals(otherPerson.medCon);
```
Comment (hweinian) [inline src/main/java/seedu/address/model/person/Person.java:143]: is there a typo? why is there a semicolon there

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 8]
```diff
@@ -35,7 +39,7 @@ public static Person[] getSamplePersons() {
                        new Gender("M"),
                        getTagSet("friends"),
                        new Priority(),
-                       getAppointmentSet("Dental:2024-12-25:1235-1300")
+                       getAppointmentSet("Dental:2024-12-25:1235-1300"), EMPTY_MEDCON
```
Comment (hweinian) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:42]: Sample data should not use EMPTY_MEDCON. Require proper sample data.

File: src/test/java/seedu/address/testutil/PersonBuilder.java [hunk 9]
```diff
@@ -147,15 +151,21 @@ public PersonBuilder withGender(String gender) {
         return this;
     }
 
+    public PersonBuilder withMedCon(String medCon) {
```
Comment (hweinian) [inline src/test/java/seedu/address/testutil/PersonBuilder.java:154]: Might be missing JavaDocs

**Review by hweinian — CHANGES_REQUESTED** [review]
Require fixes before merging

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 10]
```diff
@@ -43,7 +45,7 @@ public class AddCommandParser implements Parser<AddCommand> {
     public AddCommand parse(String args) throws ParseException {
         ArgumentMultimap argMultimap =
                 ArgumentTokenizer.tokenize(args, PREFIX_NAME, PREFIX_NRIC, PREFIX_DOB, PREFIX_GENDER,
-                        PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS, PREFIX_TAG);
+                        PREFIX_PHONE, PREFIX_EMAIL, PREFIX_ADDRESS, PREFIX_TAG, PREFIX_MEDCON);
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:48]: There is no need to include `PREFIX_MEDCON` for `AddCommand` since it is not part of the parameters for `AddCommand`.

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 11]
```diff
@@ -62,9 +64,14 @@ public AddCommand parse(String args) throws ParseException {
         Address address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
         Set<Tag> tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
         Set<Appointment> appointmentList = Collections.emptySet();
+        MedCon medCon = argMultimap.getValue(PREFIX_MEDCON).isPresent()
+                ? ParserUtil.parseMedCon(argMultimap.getValue(PREFIX_MEDCON).get())
+                : new MedCon("");
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:69]: As Medical Condition is not part of the parameters for `AddCommand`, you can initialise `medCon` to be an empty string

File: src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java (hunk 1 above)
Comment (yong-zhuo) [reply src/main/java/seedu/address/logic/parser/AddMedConCommandParser.java:30]: Yes, I second this 👍🏻

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 12]
```diff
@@ -51,6 +51,8 @@ public class JsonAdaptedPersonTest {
     private static final List<JsonAdaptedAppointment> VALID_APPOINTMENTS = BENSON.getAppointments().stream()
                                                                                  .map(JsonAdaptedAppointment::new)
                                                                                  .collect(Collectors.toList());
+    private static final String VALID_MEDCON = BENSON.getMedCon().toString();
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:54]: It would be good to create some test cases in this test class for your new Command

**Review by yong-zhuo — CHANGES_REQUESTED** [review]
There are a few issues related to `AddCommandParser` to be fixed. Do remember to create sufficient new test cases to pass CodeCov for CI

File: src/main/java/seedu/address/model/util/SampleDataUtil.java [hunk 13]
```diff
@@ -24,6 +25,9 @@
  * Contains utility methods for populating {@code AddressBook} with sample data.
  */
 public class SampleDataUtil {
+
+    public static final MedCon EMPTY_MEDCON = new MedCon("");
```
Comment (PresleyChew) [inline src/main/java/seedu/address/model/util/SampleDataUtil.java:29]: Could consider default constructor

File: src/main/java/seedu/address/model/util/SampleDataUtil.java (hunk 8 above)
Comment (PresleyChew) [reply src/main/java/seedu/address/model/util/SampleDataUtil.java:42]: I will amend too this for my priority portion, priority is default NONE for sample data. Will settle it in next PR.

File: src/main/resources/view/PersonListCard.fxml [hunk 14]
```diff
@@ -35,6 +35,7 @@
       <Label fx:id="phone" styleClass="cell_small_label" text="\$phone" />
       <Label fx:id="address" styleClass="cell_small_label" text="\$address" />
       <Label fx:id="email" styleClass="cell_small_label" text="\$email" />
+      <Label fx:id="medCon" styleClass="cell_small_label" text="\$medCon" />
```
Comment (PresleyChew) [inline src/main/resources/view/PersonListCard.fxml:38]: Smart to put medCon before priority for aesthetic purposes! 💯

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java [hunk 15]
```diff
@@ -203,20 +212,22 @@ public void toModelType_invalidAppointments_throwsIllegalValueException() {
...
 
     @Test
     public void toModelType_nullPriority_throwsIllegalValueException() {
         JsonAdaptedPerson person = new JsonAdaptedPerson(VALID_NAME, VALID_PHONE, VALID_EMAIL, VALID_ADDRESS,
-                VALID_DATE_OF_BIRTH, VALID_GENDER, VALID_NRIC, VALID_TAGS, null, VALID_APPOINTMENTS);
+                VALID_DATE_OF_BIRTH, VALID_GENDER, VALID_NRIC, VALID_TAGS, null, VALID_APPOINTMENTS,
+                VALID_MEDCON);
         String expectedMessage = String.format(MISSING_FIELD_MESSAGE_FORMAT, Priority.class.getSimpleName());
         assertThrows(IllegalValueException.class, expectedMessage, person::toModelType);
     }
```
Comment (PresleyChew) [inline src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:233]: Consider adding test case for invalid medCon and null medCone

**Review by PresleyChew — COMMENTED** [review]
Overall just some nitpicks, after fix should be ok! 😃

File: src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java (hunk 15 above)
Comment (TechRaj) [reply src/test/java/seedu/address/storage/JsonAdaptedPersonTest.java:233]: Thing is there isn't really any. Rmb we agreed even skibidi is a valid medCon haha. But I will update this later such that users can input multiple medCon soon

File: src/test/java/seedu/address/model/person/MedConTest.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+        assertTrue(medCon.equals(medCon));
+
+        // null -> returns false
+        assertFalse(medCon.equals(null));
+
+        // different types -> returns false
+        assertFalse(medCon.equals(5.0f));
+
+        // different values -> returns false
+        assertFalse(medCon.equals(new MedCon("Hypertension")));
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/model/person/MedConTest.java:32]: These test cases are designed well. Could consider creating more specific test cases to test the other methods found in MedCon in the future milestones

File: src/test/java/seedu/address/logic/parser/AddMedConCommandParserTest.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    public void parse_compulsoryFieldsMissing_failure() {
+        // Both NRIC and MedCon missing
+        assertParseFailure(parser, "", String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                AddMedConCommand.MESSAGE_USAGE));
+
+        // NRIC parameter missing
+        assertParseFailure(parser, MEDCON_DESC_AMY, String.format(MESSAGE_INVALID_COMMAND_FORMAT,
+                AddMedConCommand.MESSAGE_USAGE));
+    }
+}
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/logic/parser/AddMedConCommandParserTest.java:68]: Test cases are appropriate ⚡

File: src/test/java/seedu/address/logic/commands/AddMedConCommandTest.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+        AddMedConCommand command = new AddMedConCommand(nric, medCon);
+        assertThrows(CommandException.class, () -> command.execute(model));
+    }
+
+    @Test
+    public void execute_addMedCon_success() throws Exception {
+        // Setup data
+        Nric aliceNric = ALICE.getNric();
+        MedCon initialMedCon = new MedCon("Diabetes");
+        MedCon emptyMedCon = new MedCon(""); // Empty MedCon to represent optional state
```
Comment (yong-zhuo) [inline src/test/java/seedu/address/logic/commands/AddMedConCommandTest.java:69]: Just a nit: would be good to make the `MedCon` field in `Person` class be `Optional<MedCon>` instead of assigning an empty string, but you can consider this for the next milestone. Your current implementation works fine.

Comment (hweinian) [inline src/test/java/seedu/address/logic/commands/AddMedConCommandTest.java:69]: Just a nit: Rename all 'optional' word to something else if you do not intend to use the Optional class.

File: src/main/java/seedu/address/logic/parser/AddCommandParser.java [hunk 19]
```diff
@@ -64,9 +63,7 @@ public AddCommand parse(String args) throws ParseException {
         Address address = ParserUtil.parseAddress(argMultimap.getValue(PREFIX_ADDRESS).get());
         Set<Tag> tagList = ParserUtil.parseTags(argMultimap.getAllValues(PREFIX_TAG));
         Set<Appointment> appointmentList = Collections.emptySet();
-        MedCon medCon = argMultimap.getValue(PREFIX_MEDCON).isPresent()
-                ? ParserUtil.parseMedCon(argMultimap.getValue(PREFIX_MEDCON).get())
-                : new MedCon("");
+        MedCon medCon = new MedCon("");
```
Comment (yong-zhuo) [inline src/main/java/seedu/address/logic/parser/AddCommandParser.java:66]: Good fix 👍🏻

File: src/test/java/seedu/address/logic/parser/AddMedConCommandParserTest.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+
+import seedu.address.logic.commands.AddMedConCommand;
+import seedu.address.model.person.MedCon;
+import seedu.address.model.person.Nric;
+
+public class AddMedConCommandParserTest {
+    private final AddMedConCommandParser parser = new AddMedConCommandParser();
+
+    @Test
+    public void parse_allFieldsValid_success() {
```
Comment (hweinian) [inline src/test/java/seedu/address/logic/parser/AddMedConCommandParserTest.java:27]: Good fix

File: src/test/java/seedu/address/logic/commands/AddMedConCommandTest.java [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Create and execute a command to update the medical condition to a new value
+        AddMedConCommand updateMedConCommand = new AddMedConCommand(aliceNric, updatedMedCon);
+        CommandResult updatedCommandResult = updateMedConCommand.execute(model);
+
+        // Fetch the updated person again and verify that the medical condition was updated
+        Person secondUpdatedAlice = model.fetchPersonIfPresent(new NricMatchesPredicate(aliceNric))
+                .orElseThrow(() -> new AssertionError("Person not found"));
+        assertEquals(updatedMedCon, secondUpdatedAlice.getMedCon());
+    }
```
Comment (hweinian) [inline src/test/java/seedu/address/logic/commands/AddMedConCommandTest.java:98]: Extensive testing!

**Review by hweinian — COMMENTED** [review]
Overall LGTM

**Review by yong-zhuo — APPROVED** [review]
LGTM
